### DISCUSSION TITLE
Metrics on volume based on 2D slices in DataLoader

### DISCUSSION THREAD

**Discussion by o-fares** [Project-MONAI/tutorials#400] [Q&A]
Hello,

I'm using a segmentation framework with Monai on 2D slices extracted from 3D NIFTI volumes. I previoulsy used a Tensorflow framework, where I could use 2D slices in the batches and calculate metrics like Dice coefficient on 3D volumes (where it has more sense in my case).

I am using **DataLoader** in Monai to create batches from 2D slices (NIFTI), but it seems like the metrics can only be calculated on these 2D slices and not on the full volume.

Do you know if there could be any alternative allowing to calculate the metrics on 3D volumes?

**Comment by Nic-Ma** [comment]
Hi @o-fares ,

Thanks for your interest here.
I think you can use MONAI metrics following the example:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/metrics/metric.py#L122
Adding every batch of DataLoader, then `aggregate` for the final result.
I will try to enhance the doc-string with a full and clear example.

Thanks.

**Reply by o-fares** [reply]
Hi @Nic-Ma,

Thank you for your quick answer. 

I don't really understand how the function works. Is the aggregation done on every slice of a batch?
To be clearer, I will provide an example of what I would like to compute:

Assuming I have 10 3D volumes with 20 slices each. One batch contains 100 slices (so all the slices from 5 3D volumes). 
Currently, the metrics are calculated on the 100 slices (with a mean calculated at the end of each epoch). I would like to calculate the metrics **directly** on the 5 3D volumes (the mean on the 3D volumes has more interest than the mean on all the slices).

Thank you again for your help.

**Comment by Nic-Ma** [comment]
For your special requirement, I think you can do some thing like:
```py
dice_metric = MeanDice(...)
for data in dataloader:
    dice_metric(y_pred=data["image"], y=data["label"])
    metric = dice_metric.aggregate().item()  # result of the 100 slices?
```

Thanks.

**Reply by o-fares** [reply]
Maybe I could add a key in the data dictionary to separate the 5 patients ? The dice coefficient would have to be calculated on the original 3D volumes and not on the individual 20 slices.

Thanks again.

**Comment by Nic-Ma** [comment]
If you can separate the patients, maybe you can just compute MeanDice for every patient totally separately?

Thanks.

**Reply by o-fares** [reply]
Yes, the problem is that even if the slices are agglomerated for each patient, the MeanDice would be a mean on the 2D slices from the batches of one epoch in the training and not on the original 3D volumes. 

Thank you.

**Reply by Nic-Ma** [reply]
OK, I see. Seems your main problem is how to construct a 3D image from 2D slices?
Is it possible to do some thing like:
```py
2d_data = [LoadImage(i) for i in 2d_slices_paths] 
3d_data = torch.stack(2d_data)  # maybe you also need to switch the slice dim to the end
dice_metric(y_pred=3d_data, y=label)
metric = dice_metric.aggregate().item()
```

Thanks.
